File: src/entity/State.ts
/**
 * @main State
 */

import { Errata } from "./_Erratum"
import { PublicationRefs } from "./_PublicationRef"

/**
 * @title State
 */
export type State = {
  /**
   * An unique, increasing integer.
   * @integer
   * @minimum 1
   */
  id: number

  src: PublicationRefs

  /**
   * All translations for the entry, identified by IETF language tag (BCP47).
   * @minProperties 1
   */
  translations: {
    /**
     * @patternProperties ^[a-z]{2}-[A-Z]{2}$
     */
    [localeId: string]: Translation
  }
}

type Translation = {
  /**
   * The name of the state.
   * @minLength 1
   */
  name: string

  /**
   * The description of the state.
   * @markdown
   * @minLength 1
   */
  description: string

  errata?: Errata
}

File: Sources/OptolithDatabaseSchema/Types/State.swift
import FileDB

@Model
public struct State {
    /// The publications where you can find the entry.
    @MinItems(1)
    let src: [PublicationRef]

    /// All translations for the entry, identified by IETF language tag (BCP47).
    @Relationship(Locale.self)
    let translations: [String: Translation]

    @Embedded
    struct Translation {  // StateTranslation
        /// The state’s name.
        @MinLength(1)
        let name: String

        /// The description of the state.
        @MinLength(1)
        @Markdown
        let description: String

        /// A list of errata for the entry in the specific language.
        @MinItems(1)
        let errata: [Erratum]?
    }
}
